platform/engineering/design-docs/2020-12-14-cms-content-caching-strategy.md

```diff
@@ -0,0 +1,225 @@
+
+This document will not delve into implementation details of how various systems (builds, local development, etc.) will use the cache.
+
+### Background
+
+A large part of the VA.gov website is static content that is derived from a Drupal-based content management system (CMS). 
+Building the static pages of the site involves fetching the content either directly from the Drupal CMS or from a cache in Amazon S3 (Simple Storage Service).
+
+The CMS is hosted in the internal VA network, so the request for its content has to traverse the Trusted Internet Connection (TIC).
+- Only systems within the network have direct access to internal services like the CMS.
+- External systems, such as local development machines outside of the network, can only access internal services through a SOCKS proxy.
+- Publicly hosted services such as review instances or GitHub Actions workflows currently have no means of accessing internal services.
+
+To circumvent the need for the SOCKS proxy or a direct internal connection, there is a content cache that's regularly generated by the Jenkins build pipeline.
+- The build for each environment (dev, staging, prod) tries to pull the content to fully build the website.
+- If the Drupal content pull is successful, the pipeline caches that content in S3. Otherwise, the build uses a previous cache.
+- The cache is stored with a key derived from a combination of the environment name and a hash of the GraphQL query that's used to fetch all pages from Drupal.
+
+There have been issues with generating the cache key (department-of-veterans-affairs/va.gov-team#16401) that forced review instances to attempt to directly pull content from Drupal and fail (department-of-veterans-affairs/va.gov-team#16637).
+- Review instances are staging-like environments, with website and API integration, used by many veteran-facing services (VFS) teams to test changes within feature branches.
+- From the service interruption, it was explained that review instances are in a separate virtual private cloud (VPC) than the internal VPC that hosts internal services like the CMS.
+- In addition, because the review instances have configured an Internet gateway, they do not have clearance from VA to peer or otherwise connect with the internal VPC.
+- Since they don't have a SOCKS proxy configuration or direct access to the CMS, review instances rely entirely on the cache to build the environment.
+
+The problem with review instances prompted an overarching discussion around revising the content caching strategy.
+The goals of that change would be to prevent similar issues and to better support systems that depend on the cache.
+
+### High Level Design
+
+Content for builds on `master` are cached in S3 with a key such as `content-cache/master/content.tar.gz` instead of an ever-changing hash-based key.
+
+An Amazon Web Services (AWS) Lambda function pulls the content from Drupal and sync the cache periodically (every 5 minutes).
```

Comment:
At this time, the build can take more than 5 minutes in the CMS on a cold cache.  This would mean queries would run on top of each other.  Currently, a full build causes 100% CPU utilization on RDS.  This would essentially have our database server at 100% CPU all the time.  We already do the content build every hour, can you piggy pack off that job?